tests/framework/shim/windows_shim.cpp

```diff
@@ -0,0 +1,358 @@
+                           [handle](const D3DKMT_Adapter &adapter) { return handle == adapter.info.hAdapter; });
+    if (it == platform_shim.d3dkmt_adapters.end()) {
+        return STATUS_INVALID_PARAMETER;
+    }
+    auto &adapter = *it;
+
+    auto *reg_info = reinterpret_cast<LoaderQueryRegistryInfo *>(query_info->private_data);
+    reg_info->status = LOADER_QUERY_REGISTRY_STATUS_SUCCESS;
+    if (reg_info->output_value_size == 0) {
+        reg_info->output_value_size = it->path.size();
+    } else if (reg_info->output_value_size == it->path.size()) {
+        std::wstring path_wstr(1, L'\0');
+        path_wstr.assign(it->path.str().begin(), it->path.str().end());
+        wcscpy(&reg_info->output_string[0], path_wstr.c_str());
+    } else if (reg_info->output_value_size == it->path.size()) {
+        reg_info->status = LOADER_QUERY_REGISTRY_STATUS_BUFFER_OVERFLOW;
+    }
+
+    return STATUS_SUCCESS;
+}
+
+// clang-format off
+static CONFIGRET(WINAPI *REAL_CM_Get_Device_ID_List_SizeW)(PULONG pulLen, PCWSTR pszFilter, ULONG ulFlags) = CM_Get_Device_ID_List_SizeW;
+static CONFIGRET(WINAPI *REAL_CM_Get_Device_ID_ListW)(PCWSTR pszFilter, PZZWSTR Buffer, ULONG BufferLen, ULONG ulFlags) = CM_Get_Device_ID_ListW;
+static CONFIGRET(WINAPI *REAL_CM_Locate_DevNodeW)(PDEVINST pdnDevInst, DEVINSTID_W pDeviceID, ULONG ulFlags) =  CM_Locate_DevNodeW;
+static CONFIGRET(WINAPI *REAL_CM_Get_DevNode_Status)(PULONG pulStatus, PULONG pulProblemNumber, DEVINST dnDevInst, ULONG ulFlags) =  CM_Get_DevNode_Status;
+static CONFIGRET(WINAPI *REAL_CM_Get_Device_IDW)(DEVINST dnDevInst, PWSTR Buffer, ULONG BufferLen, ULONG ulFlags) =  CM_Get_Device_IDW;
+static CONFIGRET(WINAPI *REAL_CM_Get_Child)(PDEVINST pdnDevInst, DEVINST dnDevInst, ULONG ulFlags) =  CM_Get_Child;
+static CONFIGRET(WINAPI *REAL_CM_Get_DevNode_Registry_PropertyW)(DEVINST dnDevInst, ULONG ulProperty, PULONG pulRegDataType, PVOID Buffer, PULONG pulLength, ULONG ulFlags) =  CM_Get_DevNode_Registry_PropertyW;
+static CONFIGRET(WINAPI *REAL_CM_Get_Sibling)(PDEVINST pdnDevInst, DEVINST dnDevInst, ULONG ulFlags) = CM_Get_Sibling;
+// clang-format on
+
+CONFIGRET WINAPI SHIM_CM_Get_Device_ID_List_SizeW(PULONG pulLen, PCWSTR pszFilter, ULONG ulFlags) {
+    if (pulLen == nullptr) {
+        return CR_INVALID_POINTER;
+    }
+    *pulLen = platform_shim.CM_device_ID_list.size();
+    return CR_SUCCESS;
+}
+CONFIGRET WINAPI SHIM_CM_Get_Device_ID_ListW(PCWSTR pszFilter, PZZWSTR Buffer, ULONG BufferLen, ULONG ulFlags) {
+    if (Buffer != NULL) {
+        if (BufferLen < platform_shim.CM_device_ID_list.size()) return CR_BUFFER_SMALL;
+        for (size_t i = 0; i < BufferLen; i++) {
+            Buffer[i] = platform_shim.CM_device_ID_list[i];
+        }
+    }
+    return CR_SUCCESS;
+}
+// TODO
+CONFIGRET WINAPI SHIM_CM_Locate_DevNodeW(PDEVINST pdnDevInst, DEVINSTID_W pDeviceID, ULONG ulFlags) { return CR_FAILURE; }
```

Comment:
They are used in `loaderGetDeviceRegistryFiles()` and are todo because they are windows API function calls that get called but aren't the primary ones we want/need to test and I wasn't sure how much of these we actually want to implement. The shims exist for future consideration.

Reply:
Decision: leave it alone for now and implement the contents when the tests that use it are written